pkg/core/klinebuilder_state.go

```diff
@@ -0,0 +1,87 @@
+package core
+
+import (
+	"time"
+
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+// kLineBuilderState builds klines from trades and trades only
+type kLineBuilderState struct {
+	symbol string
+
+	// kLinesMap is a map from interval to accumulated kline of that interval
+	lastKLinesMap, kLinesMap map[types.Interval]*types.KLine
+}
+
+func NewKLineBuilderState(symbol string) *kLineBuilderState {
+	return &kLineBuilderState{
+		symbol:        symbol,
+		lastKLinesMap: make(map[types.Interval]*types.KLine),
+		kLinesMap:     make(map[types.Interval]*types.KLine),
+	}
+}
+
+// Reset will reset the kline of the given interval to a new open kline with the given start time
+// It's expected that this method is called when the kline is closed
+func (kb *kLineBuilderState) Reset(interval types.Interval, startTime types.Time) {
+	if lastKLine, ok := kb.kLinesMap[interval]; ok {
+		kb.lastKLinesMap[interval] = lastKLine
+	}
+	kb.kLinesMap[interval] = &types.KLine{
+		Symbol:    kb.symbol,
+		Interval:  interval,
+		StartTime: startTime,
+		Closed:    false,
+	}
+}
+
+func (kb *kLineBuilderState) GetKLine(interval types.Interval) (*types.KLine, bool) {
+	kline, ok := kb.kLinesMap[interval]
+	return kline, ok
+}
+
+func (kb *kLineBuilderState) GetLastKLine(interval types.Interval) (*types.KLine, bool) {
+	lastKLine, ok := kb.lastKLinesMap[interval]
+	return lastKLine, ok
+}
+
+func (kb *kLineBuilderState) AddTrade(trade types.Trade) {
```

Comment:
when AddTrade is called, you can return the updated klines to the caller, and these updated klines can be pushed to the stream

Reply:
e.g. 
```
itv1 --------|---itv2-------|---itv3
td1-td2-----|---td3-------|---td4
```
itv* is the ticker interval, not the kline interval, the ticker interval should be smaller than all the kline intervals, which is for pushing the updated kline periodically.

when td3 is received, we need to return "closed k1" and "new k2"